wgpu-core/src/indirect_validation.rs

```diff
@@ -0,0 +1,423 @@
+            .ok_or(CreateDispatchIndirectValidationPipelineError::InvalidId)?;
+
+        let dst_buffer = hub
+            .buffers
+            .write()
+            .remove(dst_buffer_id)
+            .ok_or(CreateDispatchIndirectValidationPipelineError::InvalidId)?;
+
+        let dst_bind_group = hub
+            .bind_groups
+            .write()
+            .remove(dst_bind_group_id)
+            .ok_or(CreateDispatchIndirectValidationPipelineError::InvalidId)?;
+
+        Ok(Self {
+            pipeline,
+            dst_buffer,
+            dst_bind_group,
+            src_bind_group_layout,
+        })
+    }
+
+    pub(crate) fn calculate_src_buffer_binding_size(device: &Device<A>, buffer: &Buffer<A>) -> u64 {
+        let alignment = device.limits.min_storage_buffer_offset_alignment as u64;
+
+        // We need to choose a binding size that can address all possible sets of 12 contiguous bytes in the buffer taking
+        // into account that the dynamic offset needs to be a multiple of `min_storage_buffer_offset_alignment`.
+
+        // Given the know variables: `offset`, `buffer_size`, `alignment` and the rule `offset + 12 <= buffer_size`.
+
+        // Let `chunks = floor(buffer_size / alignment)`.
+        // Let `chunk` be the interval `[0, chunks]`.
+        // Let `offset = alignment * chunk + r` where `r` is the interval [0, alignment - 4].
+        // Let `binding` be the interval `[offset, offset + 12]`.
+        // Let `aligned_offset = alignment * chunk`.
+        // Let `aligned_binding` be the interval `[aligned_offset, aligned_offset + r + 12]`.
+        // Let `aligned_binding_size = r + 12 = [12, alignment + 8]`.
+        // Let `min_aligned_binding_size = alignment + 8`.
+
+        // `min_aligned_binding_size` is the minimum binding size required to address all 12 contiguous bytes in the buffer
+        // but the last aligned_offset + min_aligned_binding_size might overflow the buffer. In order to avoid this we must
+        // pick a larger `binding_size` that satisfies: `last_aligned_offset + binding_size = buffer_size` and
+        // `binding_size >= min_aligned_binding_size`.
+
+        // Let `buffer_size = alignment * chunks + sr` where `sr` is the interval [0, alignment - 4].
+        // Let `last_aligned_offset = alignment * (chunks - u)` where `u` is the interval [0, chunks].
+        // => `binding_size = buffer_size - last_aligned_offset`
+        // => `binding_size = alignment * chunks + sr - alignment * (chunks - u)`
+        // => `binding_size = alignment * chunks + sr - alignment * chunks + alignment * u`
+        // => `binding_size = sr + alignment * u`
+        // => `min_aligned_binding_size <= sr + alignment * u`
+        // => `alignment + 8 <= sr + alignment * u`
+        // => `u` must be at least 2
+        // => `binding_size = sr + alignment * 2`
+
+        let binding_size = 2 * alignment + (buffer.size % alignment);
+        binding_size.min(buffer.size)
+    }
```

Comment:
I hope this makes sense. It took me a while to figure out and write down what the constraints were and how we come up with a binding size that satisfies them.